federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -218,40 +338,240 @@ func (ic *IngressController) deliverIngress(ingress types.NamespacedName, delay
+  If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
+*/
+func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
+	masterCluster, masterUID, err := ic.getMasterCluster()
+	clusterObj, clusterErr := conversion.NewCloner().DeepCopy(cluster) // Make a clone so that we don't clobber our input param
+	cluster, ok := clusterObj.(*federation_api.Cluster)
+	if clusterErr != nil || !ok {
+		glog.Errorf("Internal error: Failed clone cluster resource while attempting to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
+		return
+	}
+	if err == nil {
+		if masterCluster.Name != cluster.Name { // We're not the master, need to get in sync
+			cluster.ObjectMeta.Annotations[uidAnnotationKey] = masterUID
+			if _, err = ic.federatedApiClient.Federation().Clusters().Update(cluster); err != nil {
+				glog.Errorf("Failed to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
+				return
+			} else {
+				glog.V(4).Infof("Successfully added master ingress UID annotation (%q = %q) from master cluster %q to cluster %q.", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name)
+			}
+		} else {
+			glog.V(4).Infof("Cluster %q with ingress UID is already the master with annotation (%q = %q), no need to update.", cluster.Name, uidAnnotationKey, cluster.ObjectMeta.Annotations[uidAnnotationKey])
+		}
+	} else {
+		glog.V(2).Infof("No master cluster found to source an ingress UID from for cluster %q.  Attempting to elect new master cluster %q with ingress UID %q = %q", cluster.Name, cluster.Name, uidAnnotationKey, fallbackUID)
+		if fallbackUID != "" {
+			cluster.ObjectMeta.Annotations[uidAnnotationKey] = fallbackUID
+			if _, err = ic.federatedApiClient.Federation().Clusters().Update(cluster); err != nil {
+				glog.Errorf("Failed to add ingress UID annotation (%q = %q) to cluster %q. No master elected. Will try again later: %v", uidAnnotationKey, fallbackUID, cluster.Name, err)
+			} else {
+				glog.V(4).Infof("Successfully added ingress UID annotation (%q = %q) to cluster %q.", uidAnnotationKey, fallbackUID, cluster.Name)
+			}
+		} else {
+			glog.Errorf("No master cluster exists, and fallbackUID for cluster %q is invalid (%q).  This probably means that no clusters have an ingress controller configmap with key %q.  Federated Ingress currently supports clusters running Google Loadbalancer Controller (\"GLBC\")", cluster.Name, fallbackUID)
+		}
 	}
 }
 
 func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
-	glog.V(4).Infof("Reconciling ingress %q", ingress)
+	glog.V(4).Infof("Reconciling ingress %q for all clusters", ingress)
```

Comment:
I realize that this function is way too long and in need of refactoring - will do in a followup PR.